### DISCUSSION TITLE
How to declare a custom class's iterable size for pyright?

### DISCUSSION THREAD

**Discussion by btalb** [microsoft/pyright#6791] [Q&A] [answered]
In this MVP:

```python
from typing import NamedTuple


class DebugNt(NamedTuple):
    x: int
    y: int
    z: int


class DebugIt:
    def __iter__(self):
        yield from [1, 2, 3]

    def __len__(self):
        return 3


def debug_fn(x: int, y: int, z: int, b: bool = True):
    pass


debug_fn(*DebugNt(1, 2, 3))
debug_fn(*DebugIt())
```

`pyright` passes the first `debug_fn()` call but fails the second with:
```
error: Argument of type "int" cannot be assigned to parameter "b" of type "bool" in function "debug_fn"
    "int" is incompatible with "bool" (reportGeneralTypeIssues)
```

What's needed to make the second call pass linting?

**Comment by erictraut** [comment, marked as answer]
The Python static type system doesn't provide a way to specify a fixed static length for an iterable. It provides a way specify this specifically for a tuple (and implicitly for a namedtuple), but this requires significant special-casing within type checkers like pyright and mypy. It doesn't generalize to other iterable classes. This is understandable given that most interables do not have statically-defined lengths. They tend to be more dynamic.

I don't know if this is an option, but perhaps you can derive your custom class from `tuple`? That would leverage the special-case logic in type checkers.

**Reply by btalb** [reply]
Nice, I suspected there must be something manual going on in the type checker. 

I need to have a tuple where there's an assert that blocks construction with an invalid parameter set, and that pushed me away from tuples. Now that I know for sure it being a tuple is crucial, I'll try some sort of wrapping class / function that handles the assert.

Thanks @erictraut!
